File: nvmolkit/mmffOptimization.cpp
// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <boost/python.hpp>
#include <stdexcept>
#include <string>

#include "nvmolkit/boost_python_utils.h"
#include "nvmolkit/device_result_python.h"
#include "nvmolkit/mmff_python_utils.h"
#include "src/minimizer/bfgs_types.h"
#include "src/minimizer/fire_minimizer.h"
#include "src/minimizer/mmff_minimize.h"

namespace bp = boost::python;

namespace {

enum class MinimizerKind {
  BFGS,
  FIRE
};

nvMolKit::BfgsBackend parseBfgsBackend(const std::string& name) {
  if (name == "BATCHED") {
    return nvMolKit::BfgsBackend::BATCHED;
  }
  if (name == "PER_MOL" || name == "PER_MOLECULE") {
    return nvMolKit::BfgsBackend::PER_MOLECULE;
  }
  if (name == "HYBRID") {
    return nvMolKit::BfgsBackend::HYBRID;
  }
  throw std::invalid_argument("Unknown BFGS backend '" + name + "'. Expected 'BATCHED', 'PER_MOL', or 'HYBRID'.");
}

nvMolKit::FireBackend parseFireBackend(const std::string& name) {
  if (name == "BATCHED") {
    return nvMolKit::FireBackend::BATCHED;
  }
  if (name == "PER_MOL" || name == "PER_MOLECULE") {
    return nvMolKit::FireBackend::PER_MOLECULE;
  }
  if (name == "HYBRID") {
    return nvMolKit::FireBackend::HYBRID;
  }
  throw std::invalid_argument("Unknown FIRE backend '" + name + "'. Expected 'BATCHED', 'PER_MOL', or 'HYBRID'.");
}

MinimizerKind parseMinimizerKind(const std::string& name) {
  if (name == "BFGS" || name == "bfgs") {
    return MinimizerKind::BFGS;
  }
  if (name == "FIRE" || name == "fire") {
    return MinimizerKind::FIRE;
  }
  throw std::invalid_argument("Unknown minimizerKind '" + name + "'. Expected 'BFGS' or 'FIRE'.");
}

}  // namespace

BOOST_PYTHON_MODULE(_mmffOptimization) {
  bp::def(
    "MMFFOptimizeMoleculesConfs",
    +[](const bp::list&                       molecules,
        int                                   maxIters,
        const bp::list&                       propertiesList,
        const nvMolKit::BatchHardwareOptions& hardwareOptions,
        const std::string&                    backend,
        const std::string&                    minimizerKind,
        const nvMolKit::FireOptions&          fireOptions) -> bp::list {
      auto       molsVec    = nvMolKit::extractMolecules(molecules);
      const auto properties = nvMolKit::extractMMFFPropertiesList(propertiesList, static_cast<int>(molsVec.size()));
      const auto kind       = parseMinimizerKind(minimizerKind);
      const auto result     = kind == MinimizerKind::FIRE ?
                                nvMolKit::MMFF::MMFFOptimizeMoleculesConfsFire(molsVec,
                                                                           maxIters,
                                                                           fireOptions,
                                                                           properties,
                                                                           hardwareOptions,
                                                                           parseFireBackend(backend)) :
                                nvMolKit::MMFF::MMFFOptimizeMoleculesConfsBfgs(molsVec,
                                                                           maxIters,
                                                                           properties,
                                                                           hardwareOptions,
                                                                           parseBfgsBackend(backend));
      return nvMolKit::vectorOfVectorsToList(result);
    },
    (bp::arg("molecules"),
     bp::arg("maxIters")        = 200,
     bp::arg("properties")      = bp::list(),
     bp::arg("hardwareOptions") = nvMolKit::BatchHardwareOptions(),
     bp::arg("backend")         = std::string("HYBRID"),
     bp::arg("minimizerKind")   = std::string("BFGS"),
     bp::arg("fireOptions")     = nvMolKit::FireOptions()),
    "Optimize conformers for multiple molecules using MMFF force field.\n"
    "\n"
    "Args:\n"
    "    molecules: List of RDKit molecules to optimize\n"
    "    maxIters: Maximum number of optimization iterations (default: 200)\n"
    "    properties: MMFFProperties-compatible object with forcefield settings\n"
    "    hardwareOptions: BatchHardwareOptions object with hardware settings (default: default options)\n"
    "    backend: Minimizer backend: 'BATCHED', 'PER_MOL', or 'HYBRID' (default: 'HYBRID')\n"
    "    minimizerKind: 'BFGS' or 'FIRE' (default: 'BFGS')\n"
    "    fireOptions: FireOptions used when minimizerKind='FIRE'\n"
    "\n"
    "Returns:\n"
    "    List of lists of energies, where each inner list contains energies for conformers of one molecule");

  bp::def(
    "MMFFOptimizeMoleculesConfsDevice",
    +[](const bp::list&                       molecules,
        int                                   maxIters,
        const bp::list&                       propertiesList,
        const nvMolKit::BatchHardwareOptions& hardwareOptions,
        int                                   targetGpu,
        const std::string&                    backend,
        const std::string&                    minimizerKind,
        const nvMolKit::FireOptions&          fireOptions) -> bp::object {
      auto       molsVec    = nvMolKit::extractMolecules(molecules);
      const auto properties = nvMolKit::extractMMFFPropertiesList(propertiesList, static_cast<int>(molsVec.size()));
      const auto kind       = parseMinimizerKind(minimizerKind);
      auto       result     = kind == MinimizerKind::FIRE ?
                                nvMolKit::MMFF::MMFFMinimizeMoleculesConfsFire(molsVec,
                                                                     maxIters,
                                                                     fireOptions,
                                                                     properties,
                                                                     /*constraints=*/{},
                                                                     hardwareOptions,
                                                                     parseFireBackend(backend),
                                                                     nvMolKit::CoordinateOutput::DEVICE,
                                                                     targetGpu) :
                                nvMolKit::MMFF::MMFFMinimizeMoleculesConfs(molsVec,
                                                                 maxIters,
                                                                 /*gradTol=*/1e-4,
                                                                 properties,
                                                                 /*constraints=*/{},
                                                                 hardwareOptions,
                                                                 parseBfgsBackend(backend),
                                                                 nvMolKit::CoordinateOutput::DEVICE,
                                                                 targetGpu);
      if (!result.device.has_value()) {
        throw std::runtime_error("MMFFMinimizeMoleculesConfs(DEVICE) returned no device result");
      }
      return nvMolKit::buildOwningDevice3DResult(*result.device);
    },
    (bp::arg("molecules"),
     bp::arg("maxIters"),
     bp::arg("properties"),
     bp::arg("hardwareOptions"),
     bp::arg("targetGpu"),
     bp::arg("backend")       = std::string("HYBRID"),
     bp::arg("minimizerKind") = std::string("BFGS"),
     bp::arg("fireOptions")   = nvMolKit::FireOptions()),
    "Optimize conformers for multiple molecules using MMFF force field, returning device-resident "
    "results.\n"
    "\n"
    "Returns:\n"
    "    A Device3DResult carrying optimized coordinates, energies, and convergence flags on GPU.");
}

File: nvmolkit/mmffOptimization.py
# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""GPU-accelerated MMFF optimization for molecular conformers.

This module provides GPU-accelerated implementations of MMFF (Molecular Mechanics Force Field)
optimization for multiple molecules and conformers using CUDA and OpenMP.
"""

from collections.abc import Sequence
from typing import TYPE_CHECKING

from rdkit.Chem import AllChem

if TYPE_CHECKING:
    from rdkit.Chem import Mol
    from rdkit.ForceField.rdForceField import MMFFMolProperties

from nvmolkit import _mmffOptimization
from nvmolkit._mmff_bridge import default_rdkit_mmff_properties, make_internal_mmff_properties
from nvmolkit.types import HardwareOptions


def MMFFOptimizeMoleculesConfs(
    molecules: list["Mol"],
    maxIters: int = 200,
    properties: "MMFFMolProperties | Sequence[MMFFMolProperties | None] | None" = None,
    nonBondedThreshold: float | Sequence[float] = 100.0,
    ignoreInterfragInteractions: bool | Sequence[bool] = True,
    hardwareOptions: HardwareOptions | None = None,
) -> list[list[float]]:
    """Optimize conformers for multiple molecules using MMFF force field with BFGS minimization.

    This function performs GPU-accelerated MMFF optimization on multiple molecules with
    multiple conformers each. It uses CUDA for GPU acceleration and OpenMP for CPU
    parallelization to achieve high performance.

    Args:
        molecules: List of RDKit molecules to optimize. Each molecule should have
                  conformers already generated.
        maxIters: Maximum number of BFGS optimization iterations (default: 200)
        properties: RDKit ``MMFFMolProperties`` object, a per-molecule sequence
            of those objects, or ``None`` to use default MMFF94 settings.
        nonBondedThreshold: Radius threshold used to exclude long-range
            non-bonded interactions, either as a scalar or per-molecule sequence.
        ignoreInterfragInteractions: If ``True``, omit non-bonded terms between
            fragments. May also be provided as a per-molecule sequence.
        hardwareOptions: Configures CPU and GPU batching, threading, and device selection. Will attempt to use reasonable defaults if not set.

    Returns:
        List of lists of energies, where each inner list contains the optimized energies
        for all conformers of the corresponding molecule. The order matches the input
        molecule order and conformer iteration order.

    Raises:
        ValueError: If any molecules in the input list are None or lack MMFF atom types.
            ``e.args[0]`` is a summary message, ``e.args[1]`` is a dict
            with keys ``"none"`` (indices of None molecules) and ``"no_params"``
            (indices of molecules lacking MMFF atom types). Example::

                try:
                    MMFFOptimizeMoleculesConfs(mols, ...)
                except ValueError as e:
                    failed = e.args[1]
                    none_idx = failed["none"]
                    no_params_idx = failed["no_params"]
        RuntimeError: If CUDA operations fail or optimization encounters errors

    Example:
        >>> from rdkit import Chem
        >>> from rdkit.Chem import rdDistGeom
        >>> from nvmolkit.mmffOptimization import MMFFOptimizeMoleculesConfs
        >>> from nvmolkit.types import HardwareOptions
        >>>
        >>> # Load molecules and generate conformers
        >>> mol1 = Chem.AddHs(Chem.MolFromSmiles('CCO'))
        >>> mol2 = Chem.AddHs(Chem.MolFromSmiles('CCC'))
        >>> rdDistGeom.EmbedMultipleConfs(mol1, numConfs=5)
        >>> rdDistGeom.EmbedMultipleConfs(mol2, numConfs=3)
        >>>
        >>> # Set custom runtime performance options (optional)
        >>> hardware_options = HardwareOptions(batchSize=200, batchesPerGpu=4)
        >>> energies = MMFFOptimizeMoleculesConfs(
        ...     [mol1, mol2],
        ...     maxIters=500,
        ...     hardwareOptions=hardware_options
        ... )
        >>>
        >>> # energies[0] contains 5 energies for mol1's conformers
        >>> # energies[1] contains 3 energies for mol2's conformers

    Note:
        - Input molecules are modified in-place with optimized conformer coordinates
    """
    # Validate input
    if not molecules:
        return []

    none_indices = []
    no_params_indices = []
    for i, mol in enumerate(molecules):
        if mol is None:
            none_indices.append(i)
        elif not AllChem.MMFFHasAllMoleculeParams(mol):
            no_params_indices.append(i)

    if none_indices or no_params_indices:
        parts = []
        if none_indices:
            parts.append(f"None at indices {none_indices}")
        if no_params_indices:
            parts.append(f"lacking MMFF atom types at indices {no_params_indices}")
        raise ValueError(
            "; ".join(parts),
            {"none": none_indices, "no_params": no_params_indices},
        )

    def _normalize_scalar_or_list(value, name: str):
        if isinstance(value, Sequence) and not isinstance(value, (str, bytes)):
            if len(value) != len(molecules):
                raise ValueError(f"Expected {len(molecules)} values for {name}, got {len(value)}")
            return list(value)
        return [value for _ in molecules]

    def _normalize_properties(value):
        if value is None:
            return [default_rdkit_mmff_properties(mol) for mol in molecules]
        if isinstance(value, Sequence) and not hasattr(value, "SetMMFFVariant"):
            if len(value) != len(molecules):
                raise ValueError(f"Expected {len(molecules)} MMFFMolProperties objects, got {len(value)}")
            return [
                default_rdkit_mmff_properties(mol) if props is None else props for mol, props in zip(molecules, value)
            ]
        return [value for _ in molecules]

    # Call the C++ implementation
    if hardwareOptions is None:
        hardwareOptions = HardwareOptions()
    native_options = hardwareOptions._as_native()
    properties_list = _normalize_properties(properties)
    thresholds = _normalize_scalar_or_list(nonBondedThreshold, "nonBondedThreshold")
    interfrag_flags = _normalize_scalar_or_list(ignoreInterfragInteractions, "ignoreInterfragInteractions")
    native_properties = [
        make_internal_mmff_properties(
            props,
            non_bonded_threshold=float(threshold),
            ignore_interfrag_interactions=bool(ignore_interfrag),
        )
        for props, threshold, ignore_interfrag in zip(properties_list, thresholds, interfrag_flags)
    ]
    return _mmffOptimization.MMFFOptimizeMoleculesConfs(molecules, maxIters, native_properties, native_options)
